src/libraries/System.Linq/src/System/Linq/Sum.cs

```diff
@@ -49,6 +72,103 @@ public static partial class Enumerable
+            Vector<T> overflowTestVector = new(T.RotateRight(T.MultiplicativeIdentity, 1));
+
+            // Unroll the loop to sum 4 vectors per iteration
+            do
+            {
+                // Switch accumulators with each step to avoid an additional move operation
+                Vector<T> data = Vector.LoadUnsafe(ref ptr);
+                Vector<T> accumulator2 = accumulator + data;
+                Vector<T> overflowTracking = (accumulator2 ^ accumulator) & (accumulator2 ^ data);
+
+                data = Vector.LoadUnsafe(ref ptr, (nuint)Vector<T>.Count);
+                accumulator = accumulator2 + data;
+                overflowTracking |= (accumulator ^ accumulator2) & (accumulator ^ data);
+
+                data = Vector.LoadUnsafe(ref ptr, (nuint)Vector<T>.Count * 2);
+                accumulator2 = accumulator + data;
+                overflowTracking |= (accumulator2 ^ accumulator) & (accumulator2 ^ data);
+
+                data = Vector.LoadUnsafe(ref ptr, (nuint)Vector<T>.Count * 3);
+                accumulator = accumulator2 + data;
+                overflowTracking |= (accumulator ^ accumulator2) & (accumulator ^ data);
+
+                if ((overflowTracking & overflowTestVector) != Vector<T>.Zero)
+                {
+                    ThrowHelper.ThrowOverflowException();
+                }
+
+                ptr = ref Unsafe.Add(ref ptr, Vector<T>.Count * 4);
+                length -= Vector<T>.Count * 4;
+            } while (length >= Vector<T>.Count * 4);
+
+            // Process remaining vectors, if any, without unrolling
+            if (length >= Vector<T>.Count)
+            {
+                Vector<T> overflowTracking = Vector<T>.Zero;
+
+                do
+                {
+                    Vector<T> data = Vector.LoadUnsafe(ref ptr);
+                    Vector<T> accumulator2 = accumulator + data;
+                    overflowTracking |= (accumulator2 ^ accumulator) & (accumulator2 ^ data);
+                    accumulator = accumulator2;
+
+                    ptr = ref Unsafe.Add(ref ptr, Vector<T>.Count);
+                    length -= Vector<T>.Count;
+                } while (length >= Vector<T>.Count);
+
+                if ((overflowTracking & overflowTestVector) != Vector<T>.Zero)
+                {
+                    ThrowHelper.ThrowOverflowException();
+                }
+            }
+
+            // Add the elements in the vector horizontally.
+            // Vector.Sum doesn't perform overflow checking, instead add elements individually.
+            T result = T.Zero;
+            for (int i = 0; i < Vector<T>.Count; i++)
+            {
+                checked { result += accumulator[i]; }
+            }
+
+            // Add any remaining elements
+            for (int i = 0; i < length; i++)
+            {
+                checked { result += Unsafe.Add(ref ptr, i); }
+            }
```

Comment:
This could be handled by doing a backtrack and masking off already processed data.

Reply:
Do you have any suggestions on how to build a mask like that in a `Vector<T>`? So far the only way I've found is looping and calling `WithElement`, but looking at the resulting JIT of that loop I feel like just adding the elements from the span will be way better.

```asm
    L0126: vxorps ymm3, ymm3, ymm3
    L012a: vmovupd [ebp-0x2c], ymm3
    L012f: cmp eax, 8
    L0132: jae L020a
    L0138: vmovupd [ebp-0x2c], ymm2
    L013d: lea ecx, [ebp-0x2c]
    L0140: xor esi, esi
    L0142: mov [ecx+eax*4], esi
    L0145: vmovupd ymm2, [ebp-0x2c]
    L014a: inc eax
    L014b: cmp edx, eax
    L014d: jg short L0126
```

I considered a static ROS of zero elements followed by all bits set elements that could be indexed to find the mask. However, this is problematic because we're dealing with generic integers.

Reply:
Okay, I think I have a partial solution. Using left byte-shift operations on vectors will work to fill with zero elements, since we don't care if we move the elements being summed into different lanes. The problem is that this only works with `Vector128<T>` on Intel. So to use this solution we'd need to either:

A) Only implement as `Vector128<T>`, losing `long` support and running slower on Intel.
or B) Implement split `Vector128<T>` and `Vector256<T>` implementations, with lots of code duplication. We'd then only gain the backtrack advantages on short `int` arrays that fallback to `Vector128<T>`, old Intel processors, and ARM.

I don't think A is a good option. And I'm not sure B is worth it? Let me know what you think.